packages/api-client/src/main/Client.ts

```diff
@@ -150,70 +142,67 @@ class APIClient {
-        context = this.createContext(createdAccessToken.user, clientType);
-        accessToken = createdAccessToken;
-      })
-      .then(() => this.initEngine(context))
-      .then(() => this.accessTokenStore.updateToken(accessToken))
-      .then(() => context);
+  public async init(clientType: ClientType = ClientType.NONE): Promise<Context> {
+    const initialAccessToken = await this.transport.http.refreshAccessToken();
+    const context = this.createContext(initialAccessToken.user, clientType);
+
+    await this.initEngine(context);
+    this.accessTokenStore.updateToken(initialAccessToken);
+
+    return context;
   }
 
-  public login(loginData: LoginData): Promise<Context> {
-    let context: Context;
-    let accessToken: AccessTokenData;
-    let cookieResponse: AxiosResponse;
-
-    return Promise.resolve()
-      .then(() => this.context && this.logout({ignoreError: true}))
-      .then(() => this.auth.api.postLogin(loginData))
-      .then((response: AxiosResponse<any>) => {
-        cookieResponse = response;
-        accessToken = response.data;
-        context = this.createContext(accessToken.user, loginData.clientType);
-      })
-      .then(() => this.initEngine(context))
-      .then(() => retrieveCookie(cookieResponse, this.config.store))
-      .then(() => this.accessTokenStore.updateToken(accessToken))
-      .then(() => context);
+  public async login(loginData: LoginData): Promise<Context> {
+    if (this.context) {
+      await this.logout({ignoreError: true});
+    }
+
+    const cookieResponse = await this.auth.api.postLogin(loginData);
+    const accessToken = cookieResponse.data as AccessTokenData;
+
+    this.logger.info(`Saved initial access token. It will expire in "${accessToken.expires_in}" seconds.`, accessToken);
+
+    const context = this.createContext(accessToken.user, loginData.clientType);
+
+    await this.initEngine(context);
+    await retrieveCookie(cookieResponse, this.config.store);
+    await this.accessTokenStore.updateToken(accessToken);
+
+    return context;
   }
 
-  public register(userAccount: RegisterData, clientType: ClientType = ClientType.PERMANENT): Promise<Context> {
-    return (
-      Promise.resolve()
-        .then(() => this.context && this.logout({ignoreError: true}))
-        .then(() => this.auth.api.postRegister(userAccount))
-        /**
-         * Note:
-         * It's necessary to initialize the context (Client.createContext()) and the store (Client.initEngine())
-         * for saving the retrieved cookie from POST /access (Client.init()) in a Node environment.
-         */
-        .then((user: User) => this.createContext(user.id, clientType))
-        .then((context: Context) => this.initEngine(context))
-        .then(() => this.init(clientType))
-    );
+  public async register(userAccount: RegisterData, clientType: ClientType = ClientType.PERMANENT): Promise<Context> {
+    if (this.context) {
+      await this.logout({ignoreError: true});
```

Comment:
In the webapp we keep track of a [logout reason](https://github.com/wireapp/wire-webapp/blob/bd8d7d07bd91f80391d33da207071ed9013d627b/app/script/auth/route.js#L30). Maybe it's better to pass a logout reason (`register`, `login`, ...) instead of a boolean value.